Week 06

aliftf opened the pull request.

aliftf committed:
Feat: add assignment 1 - lecture 11 - week 6

aliftf committed:
Feat: add search employee dynamic criteria function

helenhash commented on Week 06/Lecture 11/Assignment 01/employee/src/main/java/com/example/employee/specification/EmployeeSpecification.java:
```diff
@@ -0,0 +1,89 @@
+import java.util.ArrayList;
+import java.util.List;
+
+public class EmployeeSpecification implements Specification<Employee> {
```
This should be in repository package

aliftf replied:
Thank you for the feedback, I've move the EmployeeSpecification to the repository package in my latest commit.

helenhash commented on Week 06/Lecture 11/Assignment 01/employee/src/main/java/com/example/employee/service/impl/EmployeeServiceImpl.java:
```diff
@@ -0,0 +1,177 @@
+        employeeDTO.setFirstName(employee.getFirstName());
+        employeeDTO.setLastName(employee.getLastName());
+        employeeDTO.setGender(employee.getGender());
+        employeeDTO.setHireDate(employee.getHireDate());
```
Can use lombok

aliftf replied:
Thankyou for the feedback, I've added the Lombok's `@Builder` and `@Data` annotations for mapping method of DTO -> entity and vice versa in the latest commit.

Also, I have moved the mapping method to the DTO/entity class correspondingly, thank you.

helenhash commented on Week 06/Lecture 11/Assignment 01/employee/src/main/java/com/example/employee/service/impl/EmployeeServiceImpl.java:
```diff
@@ -0,0 +1,177 @@
+    @Override
+    @Transactional
+    public Page<EmployeeDTO> getAllEmployees(Pageable pageable) {
+        Page<Employee> employees = employeeRepository.findAll(pageable);
```
Should notice when use findAll default not query

aliftf replied:
Thank you for the feedback, I've added `@Query` and `fetch` to avoid the N+1 query issues, the following is one of my updated code:

```java
@Query("SELECT e FROM Employee e LEFT JOIN FETCH e.salaries LEFT JOIN FETCH e.titles LEFT JOIN FETCH e.deptEmps LEFT JOIN FETCH e.deptManagers")
Page<Employee> findAll(Pageable pageable);
```

helenhash reviewed (changes requested):
It's good for using relationship between entites, but should explicitly specific the fetch type for relationship
Use lombok for build object
Mapping method should be moved to DTO/entity class correctpondingly.
Notice N+1 issue with find all

aliftf committed:
Fix: update assignment based on feedback

aliftf commented:
> It's good for using relationship between entites, but should explicitly specific the fetch type for relationship Use lombok for build object Mapping method should be moved to DTO/entity class correctpondingly. Notice N+1 issue with find all

Thank you for the feedback, 
updated commit: [commit](https://github.com/affandyfandy/java-alif/pull/10/commits/9a258922b27111843feb32690859b9a1396f4992)

I have completed the improvement of my assignment based on your feedback. I have write explicitly for specifying the fetch type for the relationship. I've using Lombok's `@Builder` and `@Data` annotations for mapping method of DTO -> entity and vice versa, and I've moved the mapping method to the DTO/entity class. Also, I added `@Query` for repository to avoid the N+1 issue. Thank you.

aliftf committed:
Feat: add fetch type
